parlai/agents/bert_qa/bert_qa.py

```diff
@@ -0,0 +1,443 @@
+    ):
+
+        start_indexes = self._get_best_indexes(start_logits, self.opt["n_best_size"])
+        end_indexes = self._get_best_indexes(end_logits, self.opt["n_best_size"])
+
+        _PrelimPrediction = collections.namedtuple(
+            "PrelimPrediction", ["start_index", "end_index", "start_logit", "end_logit"]
+        )
+        prelim_predictions = []
+
+        for start_index in start_indexes:
+            for end_index in end_indexes:
+                # We could hypothetically create invalid predictions
+                if end_index < start_index:
+                    continue
+                if start_context_position and start_index < start_context_position:
+                    # answer in the question
+                    continue
+                length = end_index - start_index + 1
+                if length > self.text_truncate:
+                    continue
+                prelim_predictions.append(
+                    _PrelimPrediction(
+                        start_index=start_index,
+                        end_index=end_index,
+                        start_logit=start_logits[start_index],
+                        end_logit=end_logits[end_index],
+                    )
+                )
+
+        prelim_predictions = sorted(
+            prelim_predictions,
+            key=lambda x: (x.start_logit + x.end_logit),
+            reverse=True,
+        )
+
+        _NbestPrediction = collections.namedtuple(  # pylint: disable=invalid-name
+            "NbestPrediction", ["text", "start_logit", "end_logit"]
+        )
+
+        nbest = []
+        for pred in prelim_predictions:
+            if len(nbest) >= self.opt["n_best_size"]:
+                break
+
+            if pred.start_index > 0:  # this is a non-null prediction
+                answer_tokens = self.dict.vec2txt(
+                    input_ids[pred.start_index:pred.end_index]
+                )
+                answer_text = self._clean_WordPieces(answer_tokens)
+                nbest.append(
+                    _NbestPrediction(
+                        text=answer_text,
+                        start_logit=pred.start_logit,
+                        end_logit=pred.end_logit,
+                    )
+                )
+
+        if len(nbest) > 0:
+            # return the best prediciton
+            return nbest[0].text
+        else:
+            # return an empty prediction
+            return ""
+
+    def report(self):
+        base = super().report()
+        m = {}
+        examples = self.metrics["examples"]
+        if examples > 0:
+            m["examples"] = examples
+            m["loss"] = self.metrics["loss"]
+            m["mean_loss"] = self.metrics["loss"] / examples
+
+        for k, v in m.items():
+            # clean up: rounds to sigfigs and converts tensors to floats
+            base[k] = round_sigfigs(v, 4)
+        return base
+
+    def train_step(self, batch):
+
+        tensors = self._vectorize(batch["observations"])
```

Comment:
these should already be available already I think

Reply:
I need custom vectors, those in the batch are not usable

Reply:
the other tensors are still being generated, which is wasting computation then and slowing down the training. should find a way to replace the other tokenization then (or do post-processing on it) instead of this approach.

(we can discuss in-person when you are back from PTO)

Reply:
we may need to make changes to torch agent to make this easier.